### PR TITLE
currency: boost retrieval speed when calling currency.NewCode()

### PR DESCRIPTION
**Pull request by shazbert**
# PR Description

* Implement map of currency.Item instead of slice
* Use upper case currency code string as key identifier 
* cleaned up some coderinos

This should have a good knock on affect to the speed of currency.NewPair functions as it was iterating over 3k items * 2 for each currency side and this was compounded across different routines.  

Fixes # (issue)

## Type of change

Please delete options that are not relevant and add an `x` in `[]` as item is complete.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How has this been tested

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration and
also consider improving test coverage whilst working on a certain feature or package.

- [ ] go test ./... -race
- [ ] golangci-lint run
- [ ] Test X

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation and regenerated documentation via the documentation tool
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally and on Github Actions/AppVeyor with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by gloriousCode — CHANGES_REQUESTED** [review]
# :horse_racing: :dash: :dash:

File: currency/code_types.go [hunk 1]
```diff
@@ -28,7 +28,7 @@ type Role uint8
 
 // BaseCodes defines a basket of bare currency codes
 type BaseCodes struct {
-	Items          []*Item
+	Items          map[string][]*Item
```
[the remark this answers is not in the document]
Comment (shazbert) [reply currency/code_types.go:31]: Yeah you are correct. This is just to maintain the state as it was before. It was mainly due to the same currency codes but two different types either fiat or cryptocurrency. At its current state there is no way to pull them out like you suggested `Code{Item: stored[0]}`. But I didn't want to go down that path for this PR just yet.

File: currency/code_types.go [hunk 2]
```diff
@@ -63,6 +63,18 @@ func (*Item) Lock() {}
 // Unlock implements the sync.Locker interface and forces a govet check nocopy
 func (*Item) Unlock() {}
 
+// copy does a full copy of item
+func (i *Item) copy() *Item {
```
[the remark this answers is not in the document]
Comment (shazbert) [reply currency/code_types.go:67]: It was just for code reduction in storage.go, I couldn't deference and then change role field, it's not really a big deal I can just remove it.

File: currency/code.go [hunk 3]
```diff
@@ -116,47 +120,60 @@ func (b *BaseCodes) GetFullCurrencyData() (File, error) {
...
-		if r != Unset {
-			b.Items[i].Role = r
-		}
-		if blockchain != "" {
-			b.Items[i].AssocChain = blockchain
-		}
-		if id != 0 {
-			b.Items[i].ID = id
+	stored, ok := b.Items[update.Symbol]
+	if ok {
```
[the remark this answers is not in the document]
Comment (shazbert) [reply currency/code.go:156]: If it fails to update, to stay in line with what was happening before we need to insert a new stored currency Item but with a different role.

[the remark this answers is not in the document]
Comment (gloriousCode) [reply currency/code.go:156]: Yeah I see now. Not worth having the append twice to minimise indenting

File: currency/code_types.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (gloriousCode) [reply currency/code_types.go]: Thanks for the removal!

File: currency/code_types.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (gloriousCode) [reply currency/code_types.go:31]: Thanks for the explanation, It'll be good for later then.

File: currency/code_test.go [hunk 4]
```diff
@@ -152,14 +152,20 @@ func (b *BaseCodes) assertRole(t *testing.T, c Code, r Role) {
...
-		if b.Items[x] != c.Item {
-			continue
-		}
-		if b.Items[x].Role != r {
-			t.Fatal("unexpected role")
+	stored, ok := b.Items[c.Item.Symbol]
+	if !ok {
+		t.Fatal("code pointer not found")
+	}
+	if ok {
```
Comment (gloriousCode) [inline currency/code_test.go:159]: rm this `if`, the above means this will always be true

**Review by gloriousCode — CHANGES_REQUESTED** [review]
Couple more small thingerinos

**Review by gloriousCode — APPROVED** [review]
tACK! Thanks for making those changes :racehorse:

**Comment by thrasher-** [comment]
tACK, quite a significant speed boost. Nice stuff!

```
PR:
goos: windows
goarch: amd64
pkg: github.com/thrasher-corp/gocryptotrader/currency
cpu: AMD Ryzen Threadripper 3970X 32-Core Processor
BenchmarkNewCode
BenchmarkNewCode-64
14170744	        77.24 ns/op	       8 B/op	       1 allocs/op

main:
goos: windows
goarch: amd64
pkg: github.com/thrasher-corp/gocryptotrader/currency
cpu: AMD Ryzen Threadripper 3970X 32-Core Processor
BenchmarkNewCode
BenchmarkNewCode-64 313797	      3364 ns/op	       8 B/op	       1 allocs/op

PR:
goos: linux
goarch: amd64
pkg: github.com/thrasher-corp/gocryptotrader/currency
cpu: AMD Ryzen Threadripper 3970X 32-Core Processor
BenchmarkNewCode
BenchmarkNewCode-64    	13845956	        84.42 ns/op	       8 B/op	       1 allocs/op

main:
goos: linux
goarch: amd64
pkg: github.com/thrasher-corp/gocryptotrader/currency
cpu: AMD Ryzen Threadripper 3970X 32-Core Processor
BenchmarkNewCode
BenchmarkNewCode-64    	  310576	      3759 ns/op	       8 B/op	       1 allocs/op

```

**Review by thrasher- — APPROVED** [review]

**Pull request merged**